### ISSUE TITLE
Files not passed as UploadedFile

### ISSUE THREAD

**Issue by NobodysNightmare** [intridea/grape#1022]
Hello,

I noticed that uploaded files are passed into my endpoints as `Hashie::Mash`, e.g.:

````ruby
#<Hashie::Mash filename="cat.jpg" head="lots-of-headers" name="file" tempfile=#<Tempfile:/tmp/RackMultipart20150601-8540-uwm6qu> type="application/jpeg">
````

This was not a big deal for me while mocking around, I just "converted" the hash into an `UploadedFile`, by instantiating my own one, the spirit being something like:

````ruby
Rack::Multipart::UploadedFile.new file[:tempfile].path, file[:type], true
````

### The moment it breaks

Trying to write specs for such an endpoint using [rack-test](https://github.com/brynary/rack-test) now starts to get difficult. [The internet](http://seejohncode.com/2012/04/29/quick-tip-testing-multipart-uploads-with-rspec/) will tell me, that I want to write something like:

````ruby
post :photo, :file => Rack::Test::UploadedFile.new(path, mime_type)
````

However, using this in a test now leads to my endpoint code (the one converting the `Mash` into an `UploadedFile`) already receiving an `UploadedFile` instead of a `Mash`.

### Questions

1. Is there a setup error on my side that prevents grape from converting the `UploadedFile` into a `Hashie::Mash` during the spec? Or is this the intended behaviour?
2. Why does the conversion from `UploadedFile` to `Hashie::Mash` happen in the first place? I tried searching through the codebase of this repo looking for `UploadedFile` and `tempfile`, but both keywords just gave me code that expects the conversion to have happened, I could not find the place of the actual conversion.

I am admittedly rather new to `Rack`. Until now it sufficed to understand the frameworks layered on top of it. So if I am missing a very basic point here, that would be an explanation.

Best regards

**Comment by dblock** [comment]
Can you take a look at https://github.com/dblock/grape-on-rack/blob/master/spec/api/upload_file_spec.rb, is this happening in that project as well? 

Sounds like you just don't have the right content-type on the upload in the test, but I am not sure.

**Comment by NobodysNightmare** [comment]
> Sounds like you just don't have the right content-type on the upload in the test, but I am not sure

I had a look at the content type:

````
0> env['CONTENT_TYPE']
=> "application/x-www-form-urlencoded"
````

url-encoded sounds definitely wrong, I will have a look in that direction. But basically the expected behaviour is, that I put an `UploadedFile` into my spec and my endpoint gets to see a `Hashie::Mash`, right?

I'll call back as soon as I know more.

**Comment by NobodysNightmare** [comment]
Arg...

Sorry for bothering you, the error was on my side. Twice! But due to inability to read I did not realize it.

**Fail 1:**

I handcrafted a `Rack::Multipart::UploadedFile` instead of a `Rack::Test::UploadedFile`. This led to `Rack::Test` not recognizing it as a file, thus simply sticking with URL encoding for everything.

After using `Rack::Test::UploadedFile` the spec works as it should.

**Fail 2:**

I did not *see* that, because I am unable to read:

````ruby
0> file
=> "#<Rack::Multipart::UploadedFile:0x0000000eea0318>"
````

If you look closely you will see that there are `"` around the class name. To be more clear:

````ruby
0> file.class
=> String
````
I was just seeing a string that looked as if it was a reference to an actual class.

So thanks for taking the time to help me and sorry for wasting yours. I should totally have seen this myself. :disappointed: 

Best regards
